File: frontend/src/lib/publicUrls.ts
// Copyright (c) 2026 PayGlue by André Nünninghoff
// Licensed under the Business Source License 1.1, see LICENSE.md

/**
 * The two public addresses this installation is reachable at.
 *
 * Both used to be the literal payglue.io hosts, spread across seven views. The
 * consequence was not cosmetic: the embed snippet a self-hosted operator copied
 * out of their own dashboard read
 * `<script src="https://api.payglue.io/paywall.js">`, so their readers' browsers
 * called our servers instead of theirs, and the webhook URL they handed their
 * payment provider pointed at us too (PG-238).
 *
 * `apiBaseUrl()` is the backend origin. It ends up in embed snippets and webhook
 * URLs, so it has to be the address the outside world can reach, which is not
 * necessarily the one the dashboard is served from.
 *
 * `appBaseUrl()` is where the dashboard itself lives, used for display and for
 * links back into it.
 *
 * Both fall back sensibly so that an install serving everything from one origin
 * needs no configuration at all.
 */

const trim = (value: string | undefined): string => (value ?? '').trim().replace(/\/+$/, '')

const CONFIGURED_API = trim(import.meta.env.VITE_PUBLIC_API_BASE_URL as string | undefined)
const CONFIGURED_APP = trim(import.meta.env.VITE_PUBLIC_APP_BASE_URL as string | undefined)
// The dashboard's own API client already has one of these. Reusing it means the
// common single-backend setup only has to be configured once.
const CLIENT_API = trim(import.meta.env.VITE_API_BASE_URL as string | undefined)

const currentOrigin = (): string => {
  if (typeof window === 'undefined' || !window.location) return ''
  return trim(window.location.origin)
}

/** Backend origin, without a trailing slash. */
export function apiBaseUrl(): string {
  return CONFIGURED_API || CLIENT_API || currentOrigin()
}

/** Dashboard origin, without a trailing slash. */
export function appBaseUrl(): string {
  return CONFIGURED_APP || currentOrigin()
}

/** Dashboard origin without the scheme, for places that show a bare address. */
export function appDisplayHost(): string {
  return appBaseUrl().replace(/^https?:\/\//, '')
}

/**
 * The address this installation answers support mail at, or an empty string.
 *
 * Empty on purpose when unset. The three places that offer a support contact
 * used to name ours, so an unrelated installation invited its own customers to
 * write to us about a product we do not run for them (PG-239). Callers hide the
 * contact rather than substitute a stand-in.
 */
export function supportEmail(): string {
  return ((import.meta.env.VITE_SUPPORT_EMAIL as string | undefined) ?? '').trim()
}

/** The `<script>` tag that embeds one of the public assets. */
export function embedScript(file: string, attributes: Record<string, string>): string {
  const attrs = Object.entries(attributes)
    .map(([key, value]) => ` ${key}="${value}"`)
    .join('')
  return `<script src="${apiBaseUrl()}/${file}"${attrs}><\/script>`
}

/** The webhook URL a payment provider has to call. */
export function webhookUrl(provider: string, tenantSlug: string, key?: string | null): string {
  const base = `${apiBaseUrl()}/webhooks/${provider}?tenant=${tenantSlug}`
  return key ? `${base}&key=${key}` : base
}

File: backend/src/payglue_backend/core/public_urls.py
# Copyright (c) 2026 PayGlue by André Nünninghoff
# Licensed under the Business Source License 1.1, see LICENSE.md
"""Where this installation is reachable from the outside.

Every link we put in an email, and every URL we hand to a payment provider, has
to point at the installation that produced it. These used to be literal
payglue.io addresses scattered across the email templates, which meant a
self-hosted install sent its own customers to our dashboard (PG-238).

Emails are the awkward case: they are sent from cron jobs, where no request
exists, so there is nothing to derive an origin from. `PUBLIC_APP_BASE_URL` has
to be configured for links to work at all. When it is not, `app_url` returns an
empty string and callers leave the link out rather than render one that goes
somewhere wrong.
"""

from urllib.parse import urlsplit

from django.conf import settings


def app_base() -> str:
    """Dashboard origin, without a trailing slash. Empty when unconfigured."""
    return (getattr(settings, "PUBLIC_APP_BASE_URL", "") or "").strip().rstrip("/")


def api_base() -> str:
    """Backend origin, without a trailing slash. Empty when unconfigured."""
    return (getattr(settings, "PUBLIC_API_BASE_URL", "") or "").strip().rstrip("/")


def app_url(path: str = "") -> str:
    """Absolute dashboard URL, or empty when the dashboard address is unknown.

    Callers must treat an empty result as "no link available" rather than
    falling back to a relative path: these end up in emails, where a relative
    path resolves against the mail client and goes nowhere.
    """
    base = app_base()
    if not base:
        return ""
    if not path:
        return base
    return f"{base}/{path.lstrip('/')}"


def app_host() -> str:
    """Dashboard hostname without scheme or port, for allowlist comparisons."""
    base = app_base()
    if not base:
        return ""
    return urlsplit(base).hostname or ""
